libvips/foreign/magick.c

```diff
@@ -109,37 +100,14 @@ magick_sniff(const unsigned char *bytes, size_t length)
 		memcmp(bytes + 4, "ftyp", 4) != 0)
 		return "TGA";
 
-	guint32 ifd_offset = 0;
-
-	/* Big-endian TIFF header
+	/* Try to search the internal magic list for a match.
 	 */
-	if (length >= 8 &&
-		bytes[0] == 'M' &&
-		bytes[1] == 'M' &&
-		bytes[2] == 0 &&
-		bytes[3] == 42)
-		ifd_offset = ((guint32) bytes[4] << 24) |
-			((guint32) bytes[5] << 16) |
-			((guint32) bytes[6] << 8) |
-			bytes[7];
-
-	/* Little-endian TIFF header
-	 */
-	if (length >= 8 &&
-		bytes[0] == 'I' &&
-		bytes[1] == 'I' &&
-		bytes[2] == 42 &&
-		bytes[3] == 0)
-		ifd_offset = ((guint32) bytes[7] << 24) |
-			((guint32) bytes[6] << 16) |
-			((guint32) bytes[5] << 8) |
-			bytes[4];
-
-	if (length >= 10 &&
-		ifd_offset >= 6 &&
-		bytes[8] == 'C' &&
-		bytes[9] == 'R')
-		return "CR2";
+	ExceptionInfo *exception = magick_acquire_exception();
+	const MagicInfo *magic_info = GetMagicInfo(bytes, length, exception);
```

Comment:
I don't think this will work with graphicsmagick -- they have:

```C
MagickExport const MagickInfo *
GetMagickInfo(const char *name,ExceptionInfo *exception)
```

Reply:
(where `name` is the name of a format, eg. `"GIF"`)

Reply:
Ah! They have:

```C
MagickExport const char *
GetImageMagick(const unsigned char *magick,const size_t length)
```

To find the format name from a string, then use `GetMagickInfo()` to get the `MagickInfo` for that format.

Reply:
Good point, I excluded this for GraphicsMagick with commit bc31f1b536b5794b86c25030d0fff2154a3f4997.

We already use `GetImageMagick()` here:
https://github.com/libvips/libvips/blob/df8b8e247ed410138fcacec2f43f0c5c16aecc37/libvips/foreign/magick.c#L663-L666

Reply:
... `GetImageMagick()` would only work for formats that define the `->magick` function (see e.g. the JPEG case [here](https://github.com/ImageMagick/ImageMagick/blob/4d18167db34ea2fd199acd99420d61192d708f06/coders/jpeg.c#L1773)). The various DNG/RAW formats doesn't use this, so you'll still need to use `GetMagicInfo()`.